docs/config/index.md

```diff
@@ -415,12 +415,14 @@ export default async ({ command, mode }) => {
 
 ### server.hmr
 
-- **Type:** `boolean | { protocol?: string, host?: string, port?: number, path?: string, timeout?: number, overlay?: boolean }`
+- **Type:** `boolean | { protocol?: string, host?: string, port?: number, path?: string, timeout?: number, overlay?: boolean, runtimePort?: string }`
 
   Disable or configure HMR connection (in cases where the HMR websocket must use a different address from the http server).
 
   Set `server.hmr.overlay` to `false` to disable the server error overlay.
 
+  Set `server.hmr.runtimePort` to any runtime code which interpreted as a port. e.g. `'window.location.port'` and `(() => window.location.port)()`. If this is set, `server.hmr.port` will be ignored.
```

Comment:
This would not work, would it? The config is being resolved on the server/in the plugin, not on the browser.

```suggestion
  Set `server.hmr.runtimePort` to the port that the client should connect to. This is useful when a connection is proxied and the external port differs from the `server.hmr.port`.
```

Reply:
`runtimePort` is useful when `port` can't be determined in advance. As mentioned  in [#3093]https://github.com/vitejs/vite/issues/3093, the port of proxy is generated randomly. `runtimePort` is `port` resolved in runtime.

Reply:
Right, I misunderstood. Edited my comment here to continue the discussion: https://github.com/vitejs/vite/pull/3115#issuecomment-826082191